feat: Beta2 release (#5689)
JetBrains/kotlin-web-site: docs/topics/whatsnew/whatsnew-eap.md

The Kotlin %kotlinEapVersion% release is out! Here are some details of this EAP release:

* **Standard library:** [Support for coroutine stack trace recovery and new features for checking equality and uniqueness of collection elements](#standard-library)
* **Kotlin/Native:** [Incremental compilation of `klib` artifacts enabled by default, new Swift export features, and the first Kotlin compiler native image](#kotlin-native)

> * **Kotlin/Wasm:** [Changes to top-level `require()` calls in `@JsFun` declarations and improved companion object initialization order](#kotlin-wasm)

Reviewer: ```suggestion
* **Kotlin/Wasm:** [Changes to top-level `require()` calls in `@JsFun` declarations, improved companion object initialization order, and support for Wasmtime in the Kotlin Gradle plugin](#kotlin-wasm)
```

For more information, see the feature's [KEEP](https://github.com/Kotlin/KEEP/blob/main/proposals/stdlib/KEEP-0461-stacktrace-recoverable.md).
We would appreciate your feedback in [YouTrack](https://youtrack.jetbrains.com/issue/KT-86595).

### New functions to check collection elements for equality and uniqueness
<primary-label ref="experimental-opt-in"/>
<secondary-label ref="standard-library"/>

Before Kotlin %kotlinEapVersion%, if you wanted to check whether collection elements were all distinct or all equal,
you had to use inefficient code patterns.

Kotlin %kotlinEapVersion% introduces experimental functions to fill this gap:

> Function

Reviewer: This table looks broken 😢

Checks
`.allDistinct()`
Every value in the collection is unique.
`.allDistinctBy()`
Every object has a unique value for the selected property.
`.allEqual()`
Every value in the collection is the same.
`.allEqualBy()`
Every object has the same value for the selected property.

You can use these functions on collections, sequences, and arrays. They compare elements using structural equality
just like other collection operations.

These functions are [Experimental](components-stability.md#stability-levels-explained) and require opt-in with the
`@OptIn(ExperimentalStdlibApi::class)` annotation or the `-opt-in=kotlin.ExperimentalStdlibApi` compiler option:

```kotlin
@OptIn(ExperimentalStdlibApi::class)
fun main() {
    data class Response(
        val participantId: String,
        val answer: String,
        val responseDate: String
    )

    val responses = listOf(
        Response("P001", "Yes", "2026-07-21"),
        Response("P002", "Maybe", "2026-07-21"),
        Response("P003", "No", "2026-07-21")
    )

>     // Check if there is consensus

Reviewer: ```suggestion
    // Checks if all participants gave the same answer
```

    println(responses.allEqualBy { it.answer })
    // false

>     // Check for duplicate participants

Reviewer: ```suggestion
    // Checks for duplicate participants
```

    println(responses.allDistinctBy { it.participantId })
    // true

>    // Check if all responses were submitted on the same date

Reviewer: ```suggestion
    // Checks if all responses were submitted on the same date
```

Reviewer: (missing space)

    println(responses.allEqualBy { it.responseDate })
    // true

    val answers = responses.map { it.answer }

>     // Check if answers are identical

Reviewer: ```suggestion
    // Checks if answers are identical
```

    println(answers.allEqual())
    // false

>     // Check if answers are distinct

Reviewer: ```suggestion
    // Checks if answers are distinct
```

that accept that interface. This is especially useful when you need to use pure Swift libraries that can't be directly
imported into Kotlin.

For example, declare a Kotlin interface and a function that accepts it:

```kotlin
// Kotlin
interface CryptoProvider {
   fun hashMD5(input: String): String
}

fun processHash(provider: CryptoProvider, input: String): String = provider.hashMD5(input)
```

On the Swift side, implement this interface using a pure Swift library and pass it back to Kotlin:

```swift
// Swift
import CryptoKit

class IosCryptoProvider: KotlinBase & CryptoProvider {
   func hashMD5(input: String) -> String {
       guard let data = input.data(using: .utf8) else { return "failed" }
       return Insecure.MD5.hash(data: data).description
   }
}

let provider = IosCryptoProvider()

// The call is dispatched to the Swift implementation
print(processHash(provider: provider, input: "Hello, world!"))
```

When Kotlin receives a Swift object, it treats it like an implementation of a regular interface, executing Swift code.

For more details on Swift export, see our [documentation](native-swift-export.md).

> ### Generated Swift.package for SwiftPM dependencies

Reviewer: ```suggestion
### Generated Package.swift for SwiftPM dependencies
```

Reviewer: or at least below we call it Package.swift 🤔

<secondary-label ref="native"/>

> When exporting an XCFramework that depends on SwiftPM packages, you must publish the resulting SwifPM package for it to

Reviewer: sanity check: 
shouldn't it be SwiftPM package? 🤔

> When exporting an XCFramework that depends on SwiftPM packages, you must publish the resulting SwifPM package for it to

Reviewer: ```suggestion
When exporting an XCFramework that depends on SwiftPM packages, you must publish the resulting SwiftPM package for it to
```

resolve correctly. To help with this, the `assembleSharedXCFramework` Gradle task now generates a `Package.swift` file
to be distributed along with the XCFramework.

For details, see the [SwiftPM export page](https://kotlinlang.org/docs/multiplatform/multiplatform-spm-export.html).

### Kotlin compiler native image
<primary-label ref="experimental-general"/>
<secondary-label ref="native"/>

Kotlin %kotlinEapVersion% features the first [Experimental](components-stability.md#stability-levels-explained) release of
the Kotlin compiler native image. The native image provides a drop-in replacement for the standard `kotlinc` command-line tool,
while offering faster startup time and higher performance.

To try out the native image, download the build from [GitHub Releases](https://github.com/JetBrains/kotlin/releases/tag/v%kotlinEapVersion%).

The native image also bundles the following compiler plugins you can use with the `-Xplugin` or `-Xcompiler-plugin` CLI options:

> * [Serialization(serialization.md)

Reviewer: ```suggestion
* [Serialization](serialization.md)
```

<secondary-label ref="wasm"/>

Kotlin/Wasm now initializes superclass companion objects before subclass companion objects, matching the JVM behavior.
Previously, the initialization could be reversed, leading to inconsistent behavior across platforms.

The update improves cross-platform consistency and reduces platform-specific differences in class initialization behavior.
It also enables correct handling of companion object initialization in deeper inheritance hierarchies, including cases
where intermediate classes don't declare companion objects.

### Support for Wasmtime in the Kotlin Gradle plugin
<secondary-label ref="wasm"/>

Kotlin %kotlinEapVersion% introduces support for [Wasmtime](https://docs.wasmtime.dev/) as a runtime for the `wasmWasi`
target in the Kotlin Gradle plugin.

Previously, the `wasmWasi` target supported only the Node.js runtime, which required a JavaScript bootstrap to run WASI
applications. With Wasmtime support, you can now run Kotlin/Wasm applications on a standalone WebAssembly runtime.

To use Wasmtime as the runtime for the `wasmWasi` target, add `wasmtime()` to your Gradle build file:

```kotlin
kotlin {
    wasmWasi {
        wasmtime()
    }
}
```

> We would appreciate your feedback in [YouTrack](https://youtrack.jetbrains.com/issue/KT-86633).

Reviewer: ```suggestion

We would appreciate your feedback in [YouTrack](https://youtrack.jetbrains.com/issue/KT-86633).
```

compiler automatically handles the bridging between Kotlin's suspend functions and native JavaScript's [async/await](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function)
model, which is useful for mixed Kotlin/TypeScript codebases.

To enable this feature, add the following compiler option to your `build.gradle.kts` file:

```kotlin
kotlin {
    js {
        compilations.all {
            compileTaskProvider.configure {
                compilerOptions {
                    freeCompilerArgs.add("-Xsuspend-lambda-exporting")
                }
            }
        }
    }
}
```

Then, mark the relevant declarations with `@JsExport`:

```kotlin
// Kotlin
@JsExport
class TaskRunner {
    suspend fun runTask(task: suspend () -> String): String {
        return task()
    }
}
```

From the TypeScript side, the suspend lambda appears as a regular async function:

```typescript
// TypeScript
import { TaskRunner } from "..."

const runner = new TaskRunner();
const result = await runner.runTask(async () => "done");
console.log(result); // "done"
```

> For more information on the `@JsExport` annotation, [our documentation](js-to-kotlin-interop.md#jsexport-annotation).

Reviewer: ```suggestion
For more information on the `@JsExport` annotation, see [`@JsExport` annotation](js-to-kotlin-interop.md#jsexport-annotation).
```
